syft/pkg/cataloger/dart/parse_pubspec_lock.go

```diff
@@ -31,13 +32,39 @@ type pubspecLockPackage struct {
 }
 
 type pubspecLockDescription struct {
+	isString bool
+	str      string
+	object   pubspecLockDescriptionObject
+}
+
+type pubspecLockDescriptionObject struct {
 	Name        string `yaml:"name" mapstructure:"name"`
 	URL         string `yaml:"url" mapstructure:"url"`
 	Path        string `yaml:"path" mapstructure:"path"`
 	Ref         string `yaml:"ref" mapstructure:"ref"`
 	ResolvedRef string `yaml:"resolved-ref" mapstructure:"resolved-ref"`
 }
 
+func (s *pubspecLockDescription) UnmarshalYAML(value *yaml.Node) error {
+	if value.Kind == yaml.ScalarNode {
+		// this is a string
+		s.isString = true
+		s.str = value.Value // Or Unmarshal again, not sure
+		return nil
+	}
+	if value.Kind == yaml.MappingNode {
+		// Unmarshal to s.myStruct
+		var t pubspecLockDescriptionObject
+		if err := value.Decode(&t); err != nil {
+			return err
+		}
+		s.isString = false
+		s.object = t
+		return nil
+	}
+	return errors.New("Unexpected type: Expected string or Description map")
+}
```

Comment:
I think you could avoid the struct changes and somewhat hard to read changes with the conditions below using an alternate implementation here. Something like (completely untested):
```go
func (p *pubspecLockDescription) UnmarshalYAML(value *yaml.Node) error {
	if value.Decode(&p.Name) == nil {
		return nil
	}
	type pld pubspecLockDescription // create a new type to avoid infinite recursion
	var p2 pld
	if err := value.Decode(&p2); err != nil {
		return err
	}
	*p = pubspecLockDescription(p2)
	return nil
}
```
An example of this technique [is here](https://github.com/spdx/tools-golang/blob/4b477e88fcb3d02aa302d976bc3768f4e34c5ca3/spdx/v2/v2_3/document.go#L84-L99).